### PR TITLE
Separated buffers

### PR DESCRIPTION
**Pull request by YuuuZhang**
Separated animation data from geometry data, and saved in different buffers.

### CODE REVIEW

File: Source/Converter/BinaryData.cs [hunk 1]
```diff
@@ -13,14 +14,20 @@ internal enum BinaryDataType
...
+        public string AnimationName { get; private set; }
         public BinaryWriter Writer { get; private set; }
 
         public BinaryData(string name)
         {
             Name = name;
             Writer = new BinaryWriter(new MemoryStream());
         }
-
+        public BinaryData(string name, string animationName)
```
Comment (bghgary) [inline Source/Converter/BinaryData.cs]: BinaryData is supposed to represent a single stream of data. This change doesn't make much sense.

File: Source/Converter/Converter.cs [hunk 2]
```diff
@@ -86,7 +86,12 @@ private T CreateInstance<T>()
         private BinaryData GetBinaryData(BinaryDataType type)
         {
             var binaryData = this.getBinaryData(type);
-            if (!binaryDataToBufferIndex.ContainsKey(binaryData))
+            if (type == BinaryDataType.Animation )
```
Comment (bghgary) [inline Source/Converter/Converter.cs]: The point of having a binaryDataToBufferIndex is to account for multiple buffers. We should add another entry to the map.

File: Source/Program.cs [hunk 3]
```diff
@@ -128,7 +129,7 @@ Manifest GenerateModels(ModelGroup modelGroup, string savePath)
                     var model = modelGroup.Models[comboIndex];
                     var filename = $"{modelGroup.Id}_{comboIndex:00}.gltf";
 
-                    using (var binaryData = new BinaryData($"{modelGroup.Id}_{comboIndex:00}.bin"))
+                    using (var binaryData = new BinaryData($"{modelGroup.Id}_{comboIndex:00}.bin", $"{modelGroup.Id}_animate_{comboIndex:00}.bin"))
```
Comment (bghgary) [inline Source/Program.cs]: We should probably add some comments to the Converter constructor. The getBinaryData function callback is intended to handle multiple buffers. The [`BinaryDataType`](https://github.com/KhronosGroup/glTF-Asset-Generator/blob/master/Source/Converter/BinaryData.cs#L6) determines which type the data is for. To support animation data in a separate buffer, the `type => binaryData` lambda passed in to the constructor needs to be something like:

```C#
var binaryData = new Dictionary<BinaryDataType, BinaryData>
{
    { BinaryDataType.Animation, binaryDataAnimation },
    { BinaryDataType.Mesh, binaryDataMesh },
    { BinaryDataType.Skin, binaryDataSkin },
};
var converter = new Converter(type => binaryData[type], model.CreateSchemaInstance);
```

The `ModelGroup` class probably needs to somehow specify that the buffer should be split. Either this is just a boolean or maybe the ModelGroup should own the function to return the binary data passed to the Converter constructor.

File: Source/Program.cs [hunk 4]
```diff
@@ -144,7 +145,9 @@ Manifest GenerateModels(ModelGroup modelGroup, string savePath)
 
                         // Create the .bin file and writes the model's data to it.
                         string dataFile = Path.Combine(modelGroupFolder, binaryData.Name);
+                        string animateDataFile = Path.Combine(modelGroupFolder, binaryData.AnimationName);
                         File.WriteAllBytes(dataFile, binaryData.Bytes);
+                        File.WriteAllBytes(animateDataFile, binaryData.Bytes);
```
Comment (bghgary) [inline Source/Program.cs]: The `binaryData` has to be two different ones or this will write the data twice.

File: Source/Program.cs (hunk 3 above)
Comment (YuuuZhang) [reply Source/Program.cs]: Add a boolean value to the ModelGroup class to specify that the buffer should be split.
Also add a function to create the binary data needed in Converter.cs.

File: Source/Converter/Converter.cs (hunk 2 above)
Comment (YuuuZhang) [reply Source/Converter/Converter.cs]: Have keep the entry to the map. Cause add a function in ModelGroup class to handle the individual buffer.

File: Source/Converter/BinaryData.cs [hunk 5]
```diff
@@ -1,8 +1,10 @@
 ﻿using System;
 using System.IO;
+using System.Collections.Generic;
```
Comment (bghgary) [inline Source/Converter/BinaryData.cs]: Should revert this file completely.

File: Source/Converter/Converter.cs [hunk 6]
```diff
@@ -57,7 +57,7 @@ public DataConvertArgs(string accessorName, string bufferViewName, Data data, bo
             }
         }
 
-        public Converter(Func<BinaryDataType, BinaryData> getBinaryData, Func<Type, object> createInstance = null)
+        public Converter(bool individual, Func<BinaryDataType, BinaryData> getBinaryData, Func<Type, object> createInstance = null)
```
Comment (bghgary) [inline Source/Converter/Converter.cs]: revert? where is individual used?

File: Source/ModelGroup.cs [hunk 7]
```diff
@@ -20,7 +21,7 @@ internal abstract partial class ModelGroup
         public List<Runtime.Image> UsedTextures { get; private set; }
         public List<Runtime.Image> UsedFigures { get; private set; }
         public bool NoSampleImages { get; protected set; }
-
+
```
Comment (bghgary) [inline Source/ModelGroup.cs]: nit: extra whitespace

File: Source/ModelGroup.cs [hunk 8]
```diff
@@ -232,6 +233,31 @@ protected void SortPropertiesList(List<Property> properties)
                 properties.Sort((x, y) => x.Name.CompareTo(y.Name));
             }
         }
+
+        /// <summary>
+        ///  Cteate glTF binary data
+        /// </summary>
+        /// <returns> binary data files </returns>
+        internal Dictionary<BinaryDataType, BinaryData> CreateBinaryData(bool individual, ModelGroupId modelGroupId, int comboIndex)
```
Comment (bghgary) [inline Source/ModelGroup.cs]: This is a weird function. I put some code in a comment in Program.cs that shows what I would do.

File: Source/ModelGroups/Buffer_Misc.cs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+namespace AssetGenerator.ModelGroups
+{
+    internal class Buffer_Misc : ModelGroup
+    {
+        public override ModelGroupId Id => ModelGroupId.Buffer_Misc;
+
+        public Buffer_Misc(List<string> imageList)
+        {
+            var baseColorTexture = new Texture { Source = UseTexture(imageList, "BaseColor_Grey")};
```
Comment (bghgary) [inline Source/ModelGroups/Buffer_Misc.cs]: Each test should ideally use the minimum set of features to test. There is no reason to include a texture in this test.

File: Source/ModelGroups/Buffer_Misc.cs [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+            // Track the common properties for use in the readme
+            CommonProperties.Add(new Property(PropertyName.BaseColorTexture, baseColorTexture.Source.ToReadmeString()));
+
+            Model CreateModel(Action<List<Property>, Runtime.MeshPrimitive> setProperties) 
+            {
+                var properties = new List<Property>();
+                Runtime.MeshPrimitive meshPrimitive = MeshPrimitive.CreateSinglePlane();
```
Comment (bghgary) [inline Source/ModelGroups/Buffer_Misc.cs]:
```suggestion
                Runtime.MeshPrimitive meshPrimitive = MeshPrimitive.CreateSinglePlane(includeTextureCoords: false);
```

You can pass in false for the `includeTextureCoords` parameter to exclude texture coordinates.

File: Source/ModelGroups/Buffer_Misc.cs [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+                        new Vector3(0.0f, 0.0f, 1.0f),
+                    }
+                );
+                meshPrimitive.Material = new Runtime.Material
+                {
+                    PbrMetallicRoughness = new PbrMetallicRoughness
+                    {
+                        BaseColorTexture = new TextureInfo { Texture = baseColorTexture },
+                    },
+                };
```
Comment (bghgary) [inline Source/ModelGroups/Buffer_Misc.cs]: There is no need for vertex colors or a material.

File: Source/ModelGroups/Buffer_Misc.cs [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+            }
+            void SetColorTypeFloat(List<Property> properties, Runtime.MeshPrimitive meshPrimitive)
+            {
+                meshPrimitive.Colors.OutputType = DataType.Float;
+                properties.Add(new Property(PropertyName.VertexColor, $"Vector3 {meshPrimitive.Colors.OutputType.ToReadmeString()}"));
+            }
+
+            Models = new List<Model>
+            {
+                CreateModel((properties, meshPrimitive) =>
```
Comment (bghgary) [inline Source/ModelGroups/Buffer_Misc.cs]: It's probably better to describe the model using a description column. The table right now doesn't really say anything.

File: Source/ModelGroup.cs [hunk 13]
```diff
@@ -246,6 +272,7 @@ internal class Model
         public Manifest.Camera Camera = null;
         public bool Animated = false;
         public bool? Loadable = true;
+        public bool Individual = false;
```
Comment (bghgary) [inline Source/ModelGroup.cs]:
```suggestion
        public bool SeparateBuffers = false;
```

I would recommend this name.

File: Source/Program.cs [hunk 14]
```diff
@@ -127,24 +128,29 @@ Manifest GenerateModels(ModelGroup modelGroup, string savePath)
...
+                    
+                    // Create the .bin file and writes the model's data to it.
+                    foreach (BinaryDataType binaryDataType in Enum.GetValues(typeof(BinaryDataType)))
+                    {                                               
+                        if (!(binaryData[binaryDataType].Bytes.Length == 0))
+                        {
+                            string binaryDataFile = Path.Combine(modelGroupFolder, binaryData[binaryDataType].Name);
+                            File.WriteAllBytes(binaryDataFile, binaryData[binaryDataType].Bytes); 
+                            binaryData[binaryDataType].Dispose();
+                        }
```
Comment (bghgary) [inline Source/Program.cs]: Suggest doing this instead:

```C#
                    void Convert(Func<BinaryDataType, BinaryData> getBinaryData)
                    {
                        // Pass the desired properties to the runtime layer, which then coverts that data into
                        // a gltf loader object, ready to create the model.
                        var converter = new Converter(model.Individual, getBinaryData, model.CreateSchemaInstance);
                        glTFLoader.Schema.Gltf gltf = converter.Convert(model.GLTF);

                        // Makes last second changes to the model that bypass the runtime layer.
                        model.PostRuntimeChanges?.Invoke(gltf);

                        // Create the .gltf file and writes the model's data to it.
                        string assetFile = Path.Combine(modelGroupFolder, filename);
                        glTFLoader.Interface.SaveModel(gltf, assetFile);
                    }

                    void WriteBins(params BinaryData[] bins)
                    {
                        foreach (var bin in bins)
                        {
                            File.WriteAllBytes(Path.Combine(modelGroupFolder, bin.Name), bin.Bytes);
                        }
                    }

                    if (model.SeparateBuffers)
                    {
                        using (var binaryData = new BinaryData($"{modelGroup.Id}_{comboIndex:00}.bin"))
                        using (var binaryDataAnimation = new BinaryData($"{modelGroup.Id}_Animation_{comboIndex:00}.bin"))
                        {
                            Convert(type => type == BinaryDataType.Animation ? binaryDataAnimation : binaryData);
                            WriteBins(binaryData, binaryDataAnimation);
                        }
                    }
                    else
                    {
                        using (var binaryData = new BinaryData($"{modelGroup.Id}_{comboIndex:00}.bin"))
                        {
                            Convert(type => binaryData);
                            WriteBins(binaryData);
                        }
                    }
```

File: Source/ModelGroups/Buffer_Misc.cs (hunk 9 above)
Comment (YuuuZhang) [reply Source/ModelGroups/Buffer_Misc.cs]: Got it, I`ll remove the texture, vertex color and material.

File: Source/Program.cs (hunk 14 above)
Comment (YuuuZhang) [reply Source/Program.cs]: Thanks for your suggestion.
I want to explain why I change Program.cs in this way : 
Because in my opinion, separate buffer is only for Buffer_Misc, so I do not deal with separate buffer in Program.cs : 
I add function  `CreateBinaryData( )` in `ModelGroup.cs` to check `SeparateBuffers `and create binary data. Then call it in `Program.cs`.

File: Source/ModelGroup.cs (hunk 8 above)
Comment (YuuuZhang) [reply Source/ModelGroup.cs]: I explain the function in the last comment.

File: Source/ReadmeTemplates/Buffer_Misc.md [hunk 15]
```diff
@@ -0,0 +1,9 @@
+These models are intended to test using an individual buffer.  
+
+The following table shows the properties that are set for every model.  
+
+~~HeaderTable~~
```
Comment (bghgary) [inline Source/ReadmeTemplates/Buffer_Misc.md]: This section is empty. Probably should remove it.

File: Source/ReadmeTemplates/Buffer_Misc.md [hunk 16]
```diff
@@ -0,0 +1,9 @@
+These models are intended to test using an individual buffer.
```
Comment (bghgary) [inline Source/ReadmeTemplates/Buffer_Misc.md]:
```suggestion
These models are intended to test edge cases for buffers.  
```

File: Source/ReadmeTemplates/Buffer_Misc.md [hunk 17]
```diff
@@ -0,0 +1,9 @@
+These models are intended to test using an individual buffer.  
+
+The following table shows the properties that are set for every model.  
+
+~~HeaderTable~~
+
+The following table shows the properties that are set for a given model.  
+
```
Comment (bghgary) [inline Source/ReadmeTemplates/Buffer_Misc.md]: Would recommend removing this text since it doesn't make sense right now.

File: Source/ModelGroups/Buffer_Misc.cs [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+                };
+            }
+
+            Models = new List<Model>
+            {
+                CreateModel((properties, channel, node) =>
+                {
+                    setTranslationChanneltarget(channel, node);
+                    SetLinearSamplerForTranslation(channel);
+                    properties.Add(new Property(PropertyName.Description, "There is one animation with one channel that targets the translation of the node"));
```
[the remark this answers is not in the document]
Comment (YuuuZhang) [reply Source/ModelGroups/Buffer_Misc.cs]: Got it. Thank you, Gary.

File: Source/Program.cs [hunk 19]
```diff
@@ -128,23 +129,46 @@ Manifest GenerateModels(ModelGroup modelGroup, string savePath)
...
+                    }
+
+                    if (!model.SeparateBuffers)
+                    {
+                        using (var binaryData = new BinaryData($"{modelGroup.Id}_{comboIndex:00}.bin"))
+                        {
+                            Convert(type => binaryData);
+                            WriteBinaryDataFiles(binaryData);
+                        }
```
[the remark this answers is not in the document]
Comment (YuuuZhang) [reply Source/Program.cs]: Got it, I`ll use VS to format the document instead of only check white space in compare changes.

File: Source/ReadmeTemplates/Buffer_Misc.md [hunk 20]
```diff
@@ -1,9 +1 @@
-These models are intended to test using an individual buffer.  
-
-The following table shows the properties that are set for every model.  
-
-~~HeaderTable~~
-
-The following table shows the properties that are set for a given model.  
-
-~~Table~~
```
Comment (bghgary) [inline Source/ReadmeTemplates/Buffer_Misc.md]: This line still need to be there or the table won't show up.
